File: apps/rivethub-web/src/lib/attachments.ts
/**
 * Composer attachment state — pure helpers so the chip lifecycle is
 * unit-testable. Files are staged through the gateway upload endpoint
 * (6h TTL, node-resolvable uri); the message references them as
 * `[attached: <uri>]` lines, which every harness reads as a path in the
 * prompt text — the one channel that works on BOTH the control-plane turn
 * path and the PTY inject path. (UserTurn.attachments exists for
 * control-plane drivers; moving bound sessions onto it needs the outbound
 * queue to carry attachments — noted follow-up, not this change.)
 */

export interface PendingAttachment {
  id: string
  name: string
  size: number
  mime: string
  status: 'uploading' | 'ready' | 'failed'
  uri?: string
}

export function markStaged(
  atts: PendingAttachment[],
  id: string,
  uri: string,
): PendingAttachment[] {
  return atts.map((a) => (a.id === id ? { ...a, status: 'ready' as const, uri } : a))
}

export function markFailed(atts: PendingAttachment[], id: string): PendingAttachment[] {
  return atts.map((a) => (a.id === id ? { ...a, status: 'failed' as const } : a))
}

export function withoutAttachment(atts: PendingAttachment[], id: string): PendingAttachment[] {
  return atts.filter((a) => a.id !== id)
}

export function anyUploading(atts: PendingAttachment[]): boolean {
  return atts.some((a) => a.status === 'uploading')
}

/** A uri interpolated into `[attached: …]` must not be able to leave the
 *  bracket line: control chars split it and `]` closes it early, smuggling
 *  arbitrary prompt text under the app's voice (the server echoes the staged
 *  file NAME into the uri, so it is not fully server-controlled). Percent-
 *  encode the closers, strip the controls. */
function sanitizeUri(uri: string): string {
  // eslint-disable-next-line no-control-regex
  return uri.replace(/[\u0000-\u001f\u007f]/g, '').replaceAll(']', '%5D')
}

/** Message text with `[attached: …]` reference lines for the staged files.
 *  Failed/uploading entries never make it into the message. */
export function withAttachmentText(text: string, atts: PendingAttachment[]): string {
  const lines = atts
    .filter((a) => a.status === 'ready' && a.uri)
    .map((a) => `[attached: ${sanitizeUri(a.uri as string)}]`)
  if (lines.length === 0) return text
  return text ? `${text}\n${lines.join('\n')}` : lines.join('\n')
}

export function formatBytes(n: number): string {
  if (n < 1024) return `${String(n)}B`
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(0)}KB`
  return `${(n / (1024 * 1024)).toFixed(1)}MB`
}

File: apps/rivethub-android/app/src/main/java/io/rivethub/app/plane/Attachments.kt
package io.rivethub.app.plane

enum class AttachmentStatus { UPLOADING, READY, FAILED }

data class PendingAttachment(
    val id: String,
    val name: String,
    val status: AttachmentStatus,
    val uri: String? = null,
    val mime: String? = null,
)

data class StagedTurnAttachment(
    val mime: String,
    val pathOrUri: String,
    val name: String? = null,
)

fun readyAttachments(atts: List<PendingAttachment>): List<StagedTurnAttachment> =
    atts.mapNotNull { a ->
        val uri = a.uri
        if (a.status != AttachmentStatus.READY || uri.isNullOrBlank()) null
        else StagedTurnAttachment(mime = a.mime ?: "application/octet-stream", pathOrUri = uri, name = a.name)
    }

fun anyFailed(atts: List<PendingAttachment>): Boolean =
    atts.any { it.status == AttachmentStatus.FAILED }

/** Guess an image mime from a file name when the provider omitted one. */
fun mimeFromName(name: String, fallback: String? = null): String? =
    when (name.substringAfterLast('.', "").lowercase()) {
        "jpg", "jpeg" -> "image/jpeg"
        "png" -> "image/png"
        "webp" -> "image/webp"
        "gif" -> "image/gif"
        else -> fallback
    }

fun anyUploading(atts: List<PendingAttachment>): Boolean =
    atts.any { it.status == AttachmentStatus.UPLOADING }

/**
 * Uploads go to the session's node (its denUrl), never the entry/view node.
 * Staging on the wrong node yields a uri the session's harness cannot read.
 */
fun uploadBaseUrl(sessionNodeDenUrl: String, entryUrl: String = ""): String {
    val session = sessionNodeDenUrl.trimEnd('/')
    if (session.isNotBlank()) return session
    return entryUrl.trimEnd('/')
}

/**
 * A uri interpolated into `[attached: …]` must not be able to leave the
 * bracket line: control chars split it and `]` closes it early. Percent-
 * encode the closers, strip the controls. Matches rivethub-web sanitizeUri.
 */
fun sanitizeUri(uri: String): String =
    uri.replace(Regex("[\\u0000-\\u001f\\u007f]"), "").replace("]", "%5D")

/**
 * Message text with `[attached: …]` reference lines for staged files.
 * Callers pass already-ready uris (uploading/failed chips never belong here).
 */
fun withAttachmentText(text: String, uris: List<String>): String {
    val lines = uris.filter { it.isNotBlank() }.map { "[attached: ${sanitizeUri(it)}]" }
    if (lines.isEmpty()) return text
    return if (text.isEmpty()) lines.joinToString("\n") else text + "\n" + lines.joinToString("\n")
}

fun readyUris(atts: List<PendingAttachment>): List<String> =
    atts.mapNotNull { a -> a.uri.takeIf { a.status == AttachmentStatus.READY && !it.isNullOrBlank() } }

/** den-server `MAX_UPLOAD_BYTES` (1 GiB). Unknown size (`size < 0`) is not refused here. */
const val MAX_UPLOAD_BYTES: Long = 1024L * 1024L * 1024L

fun uploadTooLarge(size: Long): Boolean = size >= 0 && size > MAX_UPLOAD_BYTES
